app/models/envelope.rb

```diff
@@ -54,15 +55,25 @@ class Envelope < ActiveRecord::Base
     joins(:envelope_community).where(envelope_communities: { name: community })
   end)
 
+  def self.by_resource_custom_id(field, id)
+    find_by('processed_resource @> ?', { field => id }.to_json)
```

Comment:
Please be sure there won't be performance problems with this search. And/or add an index to the system to enable this search.

Reply:
@science perhaps we could at start-up go over all the community id_field and create an index for them if it does not exists?

Reply:
It's possible we could (or already do) index the entire json structure. Postgres' GIN index is an amazing technology. But we'd have to set it up if we want to use it and I don't know the status on that..

Reply:
@science The `processed_resource` column already [has](https://github.com/CredentialEngine/CredentialRegistry/blob/master/db/migrate/20160223171632_create_envelopes.rb#L18) a GIN index. Querying by keys from that column is still an order of magnitude slower than quering by, say, `envelope_id`, but that's OK for retrieving a single resource, I believe.

Reply:
@science - the performance hit on this will be roughly double what it was before when searching by ceterms:ctid.  There should be no difference in performance between looking up by @id or by ceterms:ctid since both are indexed the same way; however, the search checks first for a matching @id, and if nothing is found then a second lookup by ceterms:ctid will be performed.  This is where the additional overhead happens, and I suspect it should be fairly negligible.

 We thought about using regex to determine whether the parameter was a URL in order to eliminate the need for two lookups, but since we can't be sure that the @id will always be a URL I'm not sure that will work.

As a side point, I have asked @excelsior to remove the prefix logic from this search since it is no longer required and has the potential to cause searches to stop working if the config gets changed.

Reply:
Maybe we should do the @id lookup first?

Reply:
@jbaird123 @excelsior @science I do not understand the point of doing a look up by two fields. We should only look up by the field that is id_field for the community in my opinion.

Reply:
@excelsior @science 

Summary of my conversation with @jbaird123 

Regarding the lookup urls alforithm:

My opinion is that if we want both urls to work:

 + http://lr-staging.learningtapestry.com/resources/ce-587b9648-bc0a-4161-8972-e7be1378edcc
 + http://lr-staging.learningtapestry.com/resources/http%3A%2F%2Flr-staging.learningtapestry.com%2Fresources%2Fce-587b9648-bc0a-4161-8972-e7be1378edcc 

the second one should be a redirect to the first one. I think CTI wants the first one to be the cannonical urls, and they wouldn't like to have dupplicates in the /resources/ namespace. So, the second one would be an alias. I am not sure if we actually need the second to work or if it would be enough to have the first one to work.

Regarding indexes performance, etc... I have an idea, although I am not sure because I am not Postgres expert. Perhaps we could add a field to the envelope table, resource_id and when a resource is created store there the value of the "id_field" field of the community, cterms:ctid in the case of CredReg. That way we could index by that collumn and be as fast as getting it by envelope_id.

Regarding removing the id_prefix, I am ok with it, as currently it will not be needed with the id_field working. Just make sure to remove it everywhere that is mentioned, I do not think it is in many places in the code. Perhaps the concept of id_prefix is more confusing for the user than id_field but it solved the problem only using @id.

Cheers!!

Reply:
@excelsior - As mentioned by @leolara  - if you could remove all references to id_prefix that would be great.  There is very low risk in removing that because that code was only added very recently so there should be very little impact if any.